### PR TITLE
DEV: Create form templates

### PR DESCRIPTION
**Pull request by keegangeorge**
**:tada: This PR adds the new structured form templates feature to Discourse** 
This new feature is the ability to create customizable structured form templates written in `YAML` format.

**:gear: To Enable:**
The feature is hidden by a site wide experimental flag. To enable the feature the site setting, `experimental_form_templates` must be toggled to `true` in the console.

**➕ Additions in this PR**
This PR specifically adds the basic setup for Form Templates. This includes:
- Creating a new database table `form_templates` to store form template information
- Creating a new `resource` route at `/admin/customize/form-templates` with its relevant `.rb` files (model, controller, etc.)
- Adding a new nav item in the Admin -> Customize navigation bar for _Templates_
- Adding client side pages for handling creating, viewing, editing, and deleting form templates.
- Adding predefined quick insert buttons to quickly append basic YAML template structure to `<AceEditor />` value (predefined YAML structure for each may change later as more of this feature is fleshed out)
- Adds specs for `form_templates` controller and model

**📷 Screenshots**
<details>
<summary>See Screenshots</summary>

image
_`/admin/customize/form-templates` (Index Route)_

image
_Model to quick view template (`admin-customize-form-template-view`)_

Screenshot 2023-02-06 at 16 05 39

`/admin/customize/form-templates/:id` _Editing a template_

Screenshot 2023-02-06 at 16 05 10

`/admin/customize/form-templates/new` _Creating a template_

</details>

### CODE REVIEW

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+      @icon="bullseye"
+      @label="admin.form_templates.quick_insert_fields.multiple_choice"
+      @title="admin.form_templates.quick_insert_fields.multiple_choice"
+      @action={{this.onInsertField}}
+      @actionParam="multiselect"
+    />
+  </div>
+
+  <div class="control-group">
+    {{! TODO(@keegan): Switch to <AceEditor /> after fixing visibility issues }}
```
Comment (keegangeorge) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:65]: Having some issues around getting `<AceEditor />` to show. It gets added to the DO< but isn't visible. With some tweaking the CSS in the console, I can make it appear, but since the CSS is generated in-line, need to do some more digging into how to fix this issue.

I want to use `<Ace Editor/>` here but I don't want it to be a blocker for building the core features, so I'm delaying this for later.

File: app/assets/javascripts/admin/addon/controllers/admin-customize-form-templates-edit.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+import Controller from "@ember/controller";
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/controllers/admin-customize-form-templates-edit.js:1]: Is this file needed?

Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/controllers/admin-customize-form-templates-edit.js:1]: Also there's a typo: `AdminCustomizeForTemplatesEdit` (missing m)

File: app/assets/javascripts/admin/addon/controllers/admin-customize-form-templates-new.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import Controller from "@ember/controller";
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/controllers/admin-customize-form-templates-new.js:1]: Is this file needed? (also same typo that the other one)

File: app/assets/javascripts/admin/addon/models/form-template.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  deleteTemplate(id) {
+    return ajax(`/admin/customize/form-templates/${id}.json`, {
+      type: "DELETE",
+    });
+  },
+
+  findAll() {
+    return ajax(`/admin/customize/form-templates.json`).then((model) => {
+      return model.form_templates.sort(
+        (a, b) => parseFloat(a.id) - parseFloat(b.id)
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/models/form-template.js:30]: Why float?

File: app/assets/javascripts/admin/addon/routes/admin-customize-form-templates-new.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import DiscourseRoute from "discourse/routes/discourse";
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/routes/admin-customize-form-templates-new.js:1]: Is this file needed?

**Comment by jjaffeux** [comment]
Looking solid :+1:

It does lack quite a lot of frontend tests IMO.

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      @actionParam="upload"
+    />
+    <DButton
+      @class="btn-flat btn-icon-text"
+      @icon="bullseye"
+      @label="admin.form_templates.quick_insert_fields.multiple_choice"
+      @title="admin.form_templates.quick_insert_fields.multiple_choice"
+      @action={{this.onInsertField}}
+      @actionParam="multiselect"
+    />
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:61]: Wonder if can't generate all of this from a list?

File: app/assets/javascripts/admin/addon/components/form-template/form.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+import I18n from "I18n";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+import { templateFormFields } from "admin/lib/template-form-fields";
+import FormTemplate from "admin/models/form-template";
+
+export default class FormTemplateForm extends Component {
+  @service router;
+  @service dialog;
+  @tracked formSubmitted = false;
+  @tracked templateContents = this.args.model?.template || "";
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.js:14]: wouldn't `templateContent` be more correct than `templateContents`?

File: app/assets/javascripts/admin/addon/components/form-template/form.js [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  isEditing = this.args.model?.id ? true : false;
+  templateName = this.args.model?.name || null;
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.js:16]: The `||  null` is probably not needed here

File: app/assets/javascripts/admin/addon/components/form-template/form.js [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            this.router.transitionTo("adminCustomizeFormTemplates.index");
+          })
+          .catch(popupAjaxError);
+      },
+    });
+  }
+
+  @action
+  onInsertField(type) {
+    const structure = templateFormFields.find(
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.js:77]: findBy

File: app/assets/javascripts/admin/addon/routes/admin-route-map.js [hunk 10]
```diff
@@ -97,6 +97,15 @@ export default function () {
             this.route("edit", { path: "/:field_name" });
           }
         );
+        this.route(
+          "adminCustomizeFormTemplates",
+          { path: "/form-templates", resetNamespace: true },
+          function () {
+            this.route("index", { path: "/" });
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/routes/admin-route-map.js:104]: This shouldn't be needed: https://guides.emberjs.com/release/routing/defining-your-routes/#toc_index-routes

File: app/assets/javascripts/admin/addon/templates/customize-form-templates-index.hbs [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+      </tbody>
+    </table>
+  {{/if}}
+
+  <DButton
+    @class="btn-primary"
+    @label="admin.form_templates.new_template"
+    @title="admin.form_templates.new_template"
+    @icon="plus"
+    @action={{action this.newTemplate}}
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/templates/customize-form-templates-index.hbs:30]: Is the "action" necessary here?

File: app/controllers/admin/form_templates_controller.rb [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    if template.update(name: params[:name], template: params[:template])
+      render_serialized(template, FormTemplateSerializer, root: "form_template")
+    else
+      render_json_error(template)
+    end
+  end
+
+  def destroy
+    template = FormTemplate.find(params[:id])
+    template.destroy
```
Comment (jjaffeux) [inline app/controllers/admin/form_templates_controller.rb:48]:
```suggestion
    template.destroy!
```

File: app/controllers/admin/form_templates_controller.rb [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    params.require(:name)
+    params.require(:template)
+
+    template = FormTemplate.new(name: params[:name], template: params[:template])
+
+    if template.save
+      render_serialized(template, FormTemplateSerializer, root: "form_template")
+    else
+      render_json_error(template)
+    end
```
Comment (jjaffeux) [inline app/controllers/admin/form_templates_controller.rb:24]: Any specific reason not to just use `create!` and rely on some default error handling?

File: app/controllers/admin/form_templates_controller.rb [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def update
+    template = FormTemplate.find(params[:id])
+
+    if template.update(name: params[:name], template: params[:template])
+      render_serialized(template, FormTemplateSerializer, root: "form_template")
+    else
+      render_json_error(template)
+    end
```
Comment (jjaffeux) [inline app/controllers/admin/form_templates_controller.rb:43]: Same question than create! for update!

File: db/migrate/20230202173641_create_form_templates.rb [hunk 15]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class CreateFormTemplates < ActiveRecord::Migration[7.0]
+  def change
+    create_table :form_templates do |t|
+      t.string :name, null: false
+      t.string :template, null: false
```
Comment (jjaffeux) [inline db/migrate/20230202173641_create_form_templates.rb:7]: you say text in annotation

Comment (jjaffeux) [reply db/migrate/20230202173641_create_form_templates.rb:7]: We should probably DB constraints on length too

File: app/assets/stylesheets/common/admin/customize.scss [hunk 16]
```diff
@@ -923,3 +923,80 @@ table.permalinks {
...
+
+  &--table {
+    .admin-list-item .action {
+      text-align: right;
+    }
+  }
+
+  .btn-primary {
+    margin-top: 1rem;
+  }
```
Comment (jjaffeux) [inline app/assets/stylesheets/common/admin/customize.scss:940]: A little bit unexpected here, maybe would deserve more precise selector

File: app/assets/javascripts/admin/addon/components/form-template/row-item.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
+import Component from "@glimmer/component";
+import { action } from "@ember/object";
+import showModal from "discourse/lib/show-modal";
+import { inject as service } from "@ember/service";
+import { ajax } from "discourse/lib/ajax";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+import I18n from "I18n";
+export default class FormTemplateRowItem extends Component {
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/row-item.js:8]: Nitpick but would serve a linebreak here :p

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    <AceEditor @content={{this.templateContents}} @mode="yaml" />
+  </div>
+
+  <div class="footer-buttons">
+    <DButton
+      @class="btn-primary"
+      @label="admin.form_templates.new_template_form.submit"
+      @title="admin.form_templates.new_template_form.submit"
+      @icon="check"
+      @action={{action this.onSubmit}}
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:74]: is action needed here?

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+      @disabled={{this.formSubmitted}}
+    />
+
+    {{#if this.isEditing}}
+      <DButton
+        @class="btn-danger"
+        @label="admin.form_templates.view_template.delete"
+        @title="admin.form_templates.view_template.delete"
+        @icon="trash-alt"
+        @action={{action this.onDelete}}
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:84]: is action needed here?

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+      />
+    {{/if}}
+
+    <DButton
+      @label="admin.form_templates.new_template_form.cancel"
+      @title="admin.form_templates.new_template_form.cancel"
+      @icon="times"
+      @action={{action this.onCancel}}
```
Comment (jjaffeux) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:92]: is action needed here?

File: app/assets/javascripts/admin/addon/components/form-template/form.js [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+    ).structure;
+
+    if (this.templateContents.length === 0) {
+      this.templateContents += structure;
+    } else {
+      this.templateContents += `\n${structure}`;
+    }
+  }
+
+  #handleErrors(e) {
```
Comment (pmusaraj) [inline app/assets/javascripts/admin/addon/components/form-template/form.js:88]: I believe you can use `popupAjaxError` instead of this helper?

File: db/migrate/20230206182727_change_template_to_text.rb [hunk 22]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class ChangeTemplateToText < ActiveRecord::Migration[7.0]
+  def change
+    change_column :form_templates, :template, :text
```
Comment (pmusaraj) [inline db/migrate/20230206182727_change_template_to_text.rb:5]: Since this isn't merged yet, it would be cleaner to skip this second migration, and add the column type in `db/migrate/20230202173641_create_form_templates.rb`. Would need some resetting in your local environment.

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+      @class="btn-flat btn-icon-text"
+      @icon="grip-lines"
+      @label="admin.form_templates.quick_insert_fields.short_answer"
+      @title="admin.form_templates.quick_insert_fields.short_answer"
+      @action={{this.onInsertField}}
+      @actionParam="input"
+    />
+    <DButton
+      @class="btn-flat btn-icon-text"
+      @icon="align-left"
```
Comment (pmusaraj) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:32]: This icon needs to be added to core (locally, I don't see it).

File: app/assets/javascripts/admin/addon/components/form-template/form.hbs [hunk 24]
```diff
@@ -0,0 +1,96 @@
+<div class="form-templates--form">
+  <div class="control-group">
+    <TextField
```
Comment (pmusaraj) [inline app/assets/javascripts/admin/addon/components/form-template/form.hbs:3]: This probably needs a label?

File: app/assets/javascripts/admin/addon/templates/modal/admin-customize-form-template-view.hbs [hunk 25]
```diff
@@ -0,0 +1,22 @@
...
+</DModalBody>
+<div class="modal-footer">
+  <DButton
+    class="btn-primary"
+    @action={{this.editTemplate}}
+    @icon="pencil-alt"
+    @label="admin.form_templates.view_template.edit"
+  />
+  <DButton
+    class="btn-danger"
```
Comment (pmusaraj) [inline app/assets/javascripts/admin/addon/templates/modal/admin-customize-form-template-view.hbs:13]: Super-minor, but I would have the "Delete" button right-aligned so that it is far from the "Save" action.

File: app/assets/stylesheets/common/admin/customize.scss (hunk 16 above)
Comment (pmusaraj) [reply app/assets/stylesheets/common/admin/customize.scss:940]: Yeah,. maybe move the margin to the table element just above.

File: public/javascripts/ace.js/1.4.13/mode-yaml.js [hunk 26]
```diff
@@ -0,0 +1,8 @@
...rn e.foldWidgets[n-1]="",e.foldWidgets[n+1]="","start"}else if(u==i&&r[i]=="#"&&o[i]=="#"&&e.getLine(n-2).search(/\S/)==-1)return e.foldWidgets[n-1]="start",e.foldWidgets[n+1]="","";return u!=-1&&u<i?e.foldWidgets[n-1]="start":e.foldWidgets[n-1]="",i<a?"start":""}}.call(o.prototype)}),ace.define("ace/mode/yaml",["require","exports","module","ace/lib/oop","ace/mode/text","ace/mode/yaml_highlight_rules","ace/mode/matching_brace_outdent","ace/mode/folding/coffee"],function(e,t,n){"use strict";var r=e("../lib/oop"),i=e("./text").Mode,s=e("./yaml_highlight_rules").YamlHighlightRules,o=e("./matching_brace_outdent").MatchingBraceOutdent,u=e("./folding/coffee").FoldMode,a=function(){this.HighlightRules=s,this.$outdent=new o,this.foldingRules=new u,this.$behaviour=this.$defaultBehaviour};r.inherits(a,i),function(){this.lineCommentStart=["#"],this.getNextLineIndent=function(e,t,n){var r=this.$getIndent(t);if(e=="start"){var i=t.match(/^.*[\{\(\[]\s*$/);i&&(r+=n)}return r},this.checkOutdent=function(e,t,n){return this.$outdent.checkOutdent(t,n)},this.autoOutdent=function(e,t,n){this.$outdent.autoOutdent(t,n)},this.$id="ace/mode/yaml"}.call(a.prototype),t.Mode=a});                (function() {
```
Comment (pmusaraj) [inline public/javascripts/ace.js/1.4.13/mode-yaml.js:1]: I think we need to add this to `javascript.rake`, see lines 255-263 there. 

It makes sense to separate this into its own PR as well, as a `DEV` addition to ACE.

**Review by pmusaraj — COMMENTED** [review]
Looks very good, I pulled this branch locally and liked what I saw. The comments are mostly cosmetic. 

One thing that we should do is mark the PR as `DEV: ` though, because this only adds the backend support for templates, once we add the ability for users to create topics with these templates, then we can add ` FEATURE` PR. (The prefix is used when we write version release notes, and therefore, it's best to hold off announcing this until we're ready.)

File: public/javascripts/ace.js/1.4.13/mode-yaml.js (hunk 26 above)
Comment (keegangeorge) [reply public/javascripts/ace.js/1.4.13/mode-yaml.js:1]: Sounds good, I'll remove `mode-yaml.js` for this PR and move this along with the addition in `javascript.rake` to a separate PR 👍🏽

File: db/migrate/20230202173641_create_form_templates.rb [hunk 27]
```diff
@@ -0,0 +1,14 @@
+# frozen_string_literal: true
+
+class CreateFormTemplates < ActiveRecord::Migration[7.0]
+  def change
+    create_table :form_templates do |t|
+      t.string :name, null: false
+      t.text :template, null: false
```
Comment (jjaffeux) [inline db/migrate/20230202173641_create_form_templates.rb:7]: We still need to enforce limits at db level here, see our lore topic

Comment (keegangeorge) [reply db/migrate/20230202173641_create_form_templates.rb]: Ah yes, forgot about that. Just pushed a commit.

File: app/assets/javascripts/admin/addon/lib/template-form-fields.js [hunk 28]
```diff
@@ -0,0 +1,69 @@
+export const templateFormFields = [
+  {
+    type: "checkbox",
+    structure: `- type: checkbox
+  choices:
+    - "Option 1"
+    - "Option 2"
+    - "Option 3"
+  attributes:
+    label: "Enter question here"
```
Comment (pmusaraj) [inline app/assets/javascripts/admin/addon/lib/template-form-fields.js:10]: It can be for later, but these strings here will need translations.

File: app/assets/stylesheets/common/admin/customize.scss [hunk 29]
```diff
@@ -923,3 +923,86 @@ table.permalinks {
...
+    display: flex;
+    align-items: center;
+    flex-wrap: wrap;
+    margin-left: 1rem;
+
+    span {
+      margin-right: 0.25rem;
+    }
+
+    .btn {
```
Comment (pmusaraj) [inline app/assets/stylesheets/common/admin/customize.scss:988]: Nitpicky, but is the button styling here necessary? Shouldn't regular buttons work just fine?

File: app/serializers/form_template_serializer.rb [hunk 30]
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+class FormTemplateSerializer < ApplicationSerializer
```
Comment (pmusaraj) [inline app/serializers/form_template_serializer.rb:3]: Do you anticipate we will use this serializer outside of admin routes? If not, then I'd recommend prepending `Admin` to the serializer name and file here.

File: config/locales/client.en.yml [hunk 31]
```diff
@@ -5507,6 +5507,41 @@ en:
...
+        new_template_form:
+          submit: "Save"
+          cancel: "Cancel"
+          name:
+            label: "Template Name"
+            placeholder: "Enter a name for this template..."
+          template:
+            label: "Template"
+            placeholder: "Create a YAML template here..."
+        delete_confirm: "Are you sure you would like to delete %{template_name} template?"
```
Comment (pmusaraj) [inline config/locales/client.en.yml:5536]: Very minor, but depending on the name of the template, this can be hard to read. I'd be fine with just "Are you sure you would like to delete this template?".

**Review by pmusaraj — COMMENTED** [review]
Getting closer and closer. My comments are minor, I think the main blocker is Joffrey's note on the constraints. 

System specs look good!

File: app/assets/javascripts/admin/addon/controllers/modals/admin-customize-form-template-view.js [hunk 32]
```diff
@@ -0,0 +1,36 @@
+import Controller from "@ember/controller";
+import ModalFunctionality from "discourse/mixins/modal-functionality";
+import { action } from "@ember/object";
+import { inject as service } from "@ember/service";
+import I18n from "I18n";
+import { popupAjaxError } from "discourse/lib/ajax-error";
+import { ajax } from "discourse/lib/ajax";
+export default class AdminCustomizeFormTemplateView extends Controller.extend(
```
[the remark this answers is not in the document]
Comment (keegangeorge) [reply app/assets/javascripts/admin/addon/controllers/modals/admin-customize-form-template-view.js:8]: Thanks for catching this, my VSCode always makes it seem like there's already a gap 😅 

Screenshot 2023-02-08 at 09 37 19

File: app/assets/stylesheets/common/admin/customize.scss (hunk 29 above)
Comment (keegangeorge) [reply app/assets/stylesheets/common/admin/customize.scss:988]: Ah yes, I don't think I need this portion
```
    &:hover,
      &:focus {
        background-color: var(--primary-low);
        color: var(--primary);
      }
```

but I think in the mockup we had borders to the right.

I'll update the css

File: app/assets/javascripts/admin/addon/lib/template-form-fields.js (hunk 28 above)
Comment (keegangeorge) [reply app/assets/javascripts/admin/addon/lib/template-form-fields.js:10]: Good catch, added a TODO comment so I don't forget.

**Review by pmusaraj — APPROVED** [review]
Approving, we can keep fine-tuning things, but this is good to merge! (Oh, there's a merge conflict that needs resolving as well.)

File: app/assets/javascripts/admin/addon/controllers/modals/admin-customize-form-template-view.js (hunk 32 above)
[the remark this answers is not in the document]
Comment (jjaffeux) [reply app/assets/javascripts/admin/addon/controllers/modals/admin-customize-form-template-view.js:9]: yeh me too :p

File: spec/system/admin_customize_form_templates_spec.rb [hunk 33]
```diff
@@ -0,0 +1,141 @@
+# frozen_string_literal: true
```
Comment (jjaffeux) [inline spec/system/admin_customize_form_templates_spec.rb:1]: A high level note on specs here but I would recommend to organize things like this:

```
setup

action

expectation
```

It makes for a much easier to parse spec.

File: spec/system/admin_customize_form_templates_spec.rb [hunk 34]
```diff
@@ -0,0 +1,141 @@
...
+      form_template_page.click_view_form_template
+      expect(form_template_page).to have_template_structure("some yaml template: value")
+    end
+  end
+
+  describe "when visiting the page to edit a form template" do
+    it "should prefill form data" do
+      visit("/admin/customize/form-templates/#{form_template.id}")
+      expect(form_template_page).to have_name_value(form_template.name)
+      # difficult to test the ace editor content (todo later)
```
Comment (martin-brennan) [inline spec/system/admin_customize_form_templates_spec.rb:32]: @keegangeorge can you please open a minor PR to add your name to this TODO? We should try not to add TODO comments to the codebase without doing this since otherwise it has no ownership (I see you already did it for other TODOs in this PR which is great :+1: )
